### ISSUE TITLE
Release concept of the simh project

### ISSUE THREAD

**Issue by wfjm** [simh/simh#734]
Under [Releases](https://github.com/simh/simh/releases) I see a `v4.0-devel` dated November 21st 2014. Under [Tags](https://github.com/simh/simh/tags) I see an V3.10 dated April 16th 2016 which never made in into a release. In [Branches](https://github.com/simh/simh/branches) I see quite a few branches, but all stale not not active since 8 or more years.

When I look at Ubuntu I see
```
    xenial (16.04LTS)  3.8.1-5build1  
    bionic (18.04LTS)  3.8.1-6  
    cosmic (18.10)     3.8.1-6
    disco (19.04)      3.8.1-6
    eoan               3.8.1-6
```
Even the latest Ubuntu uses V3.8.1, which was released in February 2009. A version more than **10 years old**. I don't quite understand why Debian (and by extension Ubuntu) didn't switch to v3.9-0, the last release, but it is clear that package maintainers will not simply grab  master/HEAD and hope for the best.

So I wonder what the release concept of the simh project is.

**Comment by markpizz** [comment]
See the discussion at #377

**Comment by Rhialto** [comment]
Did you see the FOSDEM 2018 talk titled "How To Make Package Managers Cry"? A video recording is available at https://archive.fosdem.org/2018/schedule/event/how_to_make_package_managers_cry/ .

**Comment by wfjm** [comment]
Dear @markpizz,

you answered
> See the discussion at #377

The exchange on #377 is indeed telling. This was 2017, apparently in the freeze phase of [Debian 9](https://en.wikipedia.org/wiki/Debian_version_history#Debian_9_(Stretch)). When looking at the [status of the simh package in Debian](https://packages.debian.org/search?keywords=simh&searchon=names&suite=all&section=all) you see
```
Debian  8 jessie   Apr 2015: 3.8.1-5
Debian  9 stretch  Jun 2017: 3.8.1-5+b2
Debian 10 buster   Jul 2019: 3.8.1-6:
          bullseye testing : 3.8.1-6+b1
```
So Debian 9 was released with 3.8, and 2 years later Debian 10 too.
Your answer apparently frustrated the Debian maintainer @andreasbombe.
And the reasons can be most likely read from the nice FOSDEM talk ["How to make package managers cry"](https://archive.fosdem.org/2018/schedule/event/how_to_make_package_managers_cry/) posted by @Rhialto .

You wrote in #377 

> formal releases are not a priority since now that we're on github, 
> anyone can always get the best code at any time.

Imho this logic is simply not correct: What is _the best code_ and how can I figure that out _at any time_ ? 

What is _best code_ depends on context. For developers this is HEAD. But a release maintainer like @andreasbombe or somebody who just wants stable operation looks usually for the last release. Here is why: simply because releases usually are more thoroughly tested. So releases are the mechanism which allows _at any time_ to find out what is _best code_.

Sure I know about CI/CD. And I know that simh has setup Travis CI. But looking at [.travis.yml](https://github.com/simh/simh/blob/master/.travis.yml) shows that CI only checks that the simulators compile and link, and not that they work properly. This is far more involved, and any in depth verification will have to boot the usual operating systems for a given CPU. Not a simple task. And I know that I'm talking about, I'm currently setting up such an automatized testing for the [w11](https://github.com/wfjm/w11) project (testing a FPGA implementation or a software emulation CPU are very related topics). Doing such in depth testing can hardly be done per commit, can hardly be done with Travis CI (already for license reasons).

Beyond that, releases also structure the CHANGELOG.

Bottom line is that I'm strongly in favor of releases, also for projects like simh.

With best regards,   Walter

**Comment by markpizz** [comment]
> you answered
> See the discussion at #377
> The exchange on #377 is indeed telling. This was 2017, apparently in the freeze 
> phase of Debian 9. When looking at the status of the simh package in Debian you see
>
>             Debian  8 jessie   Apr 2015: 3.8.1-5
>             Debian  9 stretch  Jun 2017: 3.8.1-5+b2
>             Debian 10 buster   Jul 2019: 3.8.1-6:
>                               bullseye testing : 3.8.1-6+b1
>
> So Debian 9 was released with 3.8, and 2 years later Debian 10 too.
> Your answer apparently frustrated the Debian maintainer @andreasbombe.

I'm not sure it was frustrating or merely the fact that he started the discussion 
and looked at what was needed about 1 week before the final freeze for Debian 9.

> And the reasons can be most likely read from the nice FOSDEM talk 
> "How to make package managers cry" posted by @Rhialto .

Since, as you observe, Debian (and other distros) have been packaging 3.8.1 
forever (it was "released" on Feb 8, 2008) and no one was ever interested 
in 3.9 which was "released" on May 5, 2012, and those maintaining packages 
NEVER managed to update to that.  Given this fundamental lack of interest 
from the packaging side, the point of trying to produce formal releases to 
benefit the packaging folks would be a waste of effort.

If someone actually wanted to update the packages that were being 
distributed, I've been open to working with them to make that happen 
as I was with @andreasbombe.

> You wrote in #377
>
>             formal releases are not a priority since now that we're on github,
>             anyone can always get the best code at any time.
>
> Imho this logic is simply not correct: What is the best code and how can 
> I figure that out at any time ?
>
> What is best code depends on context. For developers this is HEAD. 

Actually it is more than developers.  If any simulator user comes and 
wants to ask questions or raise issues with the details of simh, the only 
support I'll be interested in providing relate to someone who is running 
the latest version from github.  Sure, they can as generic questions on 
the simh mailing list which others can answer, but any real problem 
will need to be raised against the github codebase.

> But a release maintainer like @andreasbombe or somebody who 
> just wants stable operation looks usually for the last release. 

Once again he walked into the maintaining activity at the last minute 
and certainly hadn't done anything before walking up at that time.

> Here is why: simply because releases usually are more thoroughly 
> tested.  So releases are the mechanism which allows at any time to 
> find out what is best code.
>
> Sure I know about CI/CD. And I know that simh has setup Travis CI. 
> But looking at .travis.yml shows that CI only checks that the simulators 
> compile and link, and not that they work properly. 

Actually, the makefile not only does a compile and link, but it also runs 
any available tests that each simulator provides to validate that 
functionality has been maintained.   The large number of VAX simulators 
provide various test code to run original diagnostics that absolutely have 
discovered problems due to various issues with compiler optimizations 
on different systems.  Many of the older simulators don't have current 
test code, but I'd be glad to accept contributions from anyone who 
wants to put tests together.  Folks doing newer simulator development 
have been implementing testing according to the needs of the developer 
writing each particular simulator.

> This is far more involved, and any in depth verification will have to boot 
> the usual operating systems for a given CPU. Not a simple task. 

Absolutely, and merely booting one of the potentially many operating 
systems is not a complete test by any means.  VAX simulators have been 
known to successfully boot a number operating systems and still fail
the diagnostic tests.

> And I know that I'm talking about, I'm currently setting up such an 
> automatized testing for the w11 project (testing a FPGA implementation 
> or a software emulation CPU are very related topics). Doing such in 
> depth testing can hardly be done per commit, can hardly be done with 
> Travis CI (already for license reasons).
>
> Beyond that, releases also structure the CHANGELOG.
> Bottom line is that I'm strongly in favor of releases, also for projects like simh.

What's been going on now has been working well for many years.  

My prior comment (in #377) specifically said "formal releases are not a priority".  By saying that, I was very specifically referring to me.  I meant: "formal releases are not a priority for me".

If they are a priority for you or others, please stand up and contribute the effort, I'd 
certainly be open to it.

**Comment by pgalbraith** [comment]
@markpizz any chance you'd be willing to slap a 4.0.0 tag on a recent commit, just to put a stick in the ground for those of us considering packing up binaries for end users?

**Comment by markpizz** [comment]
I would certainly be willing, but I'd like to see what was actually being put together in the package.  I'd hold off on a formal tag until any tweaks needed by the packaging process has been completed.

The prior packaging that I'd seen seemed to have cobbled together both the simh simulators and the simtools things.  Many of the commonly needed simtools pieces have been either integrated directly into the simh framework and are thus not needed or may only be needed by a limited domain of users or were one time use tools that a simulator developer needed while developing their simulator(s).  

Throwing everything from simh (and simtools) into a single package is one way of doing this, but all of the pieces most likely really aren't interesting to most users.  

The simtools binaries have been loosely maintained without anyone being careful about cleanly managing compile warnings and errors and really is a use at your own risk kind of thing.

Please be specific about where do you want to go with this.

**Comment by pgalbraith** [comment]
I was just intending to put Bob Supnik's 3.9 exe package, from the simh website, into a [Chocolatey](https://www.chocolatey.org) package.  When I scratched a little deeper I ended up here, and was thinking it might make sense to provide some more up to date binaries.  I hadn't even considered the tools as my usage doesn't go that deep, but off the top of my head I'd think they deserve separate packaging for niche users who need them, just as they are separated out on Bob's SimH pages.

It sounds like you might have a more comprehensive vision in the back of your mind ... If you do, I'm listening :-)

**Comment by markpizz** [comment]
We haven't really had requests for Windows releases since the latest simulators based on this github.com/simh/simh repo have been available for about 8 years at https://github.com/simh/Win32-Development-Binaries.  

This discussion has historically been about Linux releases which have a packaged simh release based on simh v3.8-1 which was released 11 years ago.  simh v3.9 was released some 8 years ago and no one in the Linux world bothered to update to that, so I thought you were coming along to pick up that ball.

Whoever originally put the Linux package together gathered up everything simh (which included simtools) and packaged it together.  That is why I asked about what you were thinking.  As a practical matter, nothing in the tools repo is really needed by the vast majority of folks who want to run a simulator.  For years, Bob provided windows binaries before the 3.9 release when his compiled binaries would run essentially everywhere.  As newer Windows versions have come along over the last 11 years, that was not in his wheel house.  The Win32-Development-Binaries will run on all versions of Windows from XP onward.
